## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-zoned.sh | :x: [more...](#e12f3042-9776-4946-b5de-2361f064b022) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#677c0dfa-7c39-4c40-90c9-c46b9f6ffbf9) |
<a name="677c0dfa-7c39-4c40-90c9-c46b9f6ffbf9"></a>
### 12213-8e9a73f/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:32399
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 695565c94ecc:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... FAILED in 32 seconds
{
  "name": "awscli",
  "duration": 2540,
  "function": "aws --endpoint-url http://minio-dev6.minio.io:32399 s3api copy-object --bucket awscli-mint-test-bucket-5685 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-5685/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 47 seconds
(7/15) Running minio-dotnet tests ... done in 38 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 15 seconds
{
  "args": {},
  "duration": 706,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 2 minutes and 21 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 130,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-2f13c1o, objectName = minio-java-test-1dh4nmm-copy, resource = /minio-java-test-2f13c1o/minio-java-test-1dh4nmm-copy, requestId = 167BDFAB3F182BCF, hostId = c2d52341-e0d0-4dbc-a059-e6df3b93cbb3)\nrequest={method=PUT, url=http://minio-dev6.minio.io:32399/minio-java-test-2f13c1o/minio-java-test-1dh4nmm-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-278rgsu/minio-java-test-1dh4nmm\nHost: minio-dev6.minio.io:32399\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210504T132017Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210504/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 418\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Tue, 04 May 2021 13:20:17 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 167BDFAB3F182BCF\nX-Xss-Protection: 1; mode=block\nDate: Tue, 04 May 2021 13:20:17 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 46 seconds
(9/15) Running minio-py tests ... done in 2 minutes and 38 seconds
(10/15) Running s3cmd tests ... FAILED in 5 seconds
{
  "name": "s3cmd",
  "duration": "2738",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-12475/s3cmd-test-object-30291 48577086-e8b8-4c88-bcf2-ef4f0aa65a33\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 48577086-e8b8-4c88-bcf2-ef4f0aa65a33 put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 5 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
<a name="e12f3042-9776-4946-b5de-2361f064b022"></a>
### 12213-8e9a73f/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:32340
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a8a6dc89ef91:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... done in 2 minutes and 16 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 1 minutes and 13 seconds
(8/15) Running minio-dotnet tests ... done in 50 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 59 seconds
(10/15) Running minio-java tests ... done in 1 minutes and 50 seconds
(11/15) Running minio-js tests ... done in 46 seconds
(12/15) Running minio-py tests ... FAILED in 2 minutes and 45 seconds
{
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-8689e5eb-04dd-4cba-88cd-48d0a41f3752",
    "object_name": "1f249d32-b1a3-4265-8890-63e9551f8536"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-8689e5eb-04dd-4cba-88cd-48d0a41f3752, request_id: 167BDFF29CB8DB9B, host_id: f5d4b01f-409c-469c-b315-3ac35020fb26, bucket_name: minio-py-test-8689e5eb-04dd-4cba-88cd-48d0a41f3752",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1640, in test_thread_safe\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-8689e5eb-04dd-4cba-88cd-48d0a41f3752, request_id: 167BDFF29CB8DB9B, host_id: f5d4b01f-409c-469c-b315-3ac35020fb26, bucket_name: minio-py-test-8689e5eb-04dd-4cba-88cd-48d0a41f3752\n",
  "duration": 6241
}
(12/15) Running s3cmd tests ... done in 17 seconds
(13/15) Running s3select tests ... done in 5 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/pkg/dsync/drwmutex.go
+++ b/pkg/dsync/drwmutex.go
@@ -49,7 +49,7 @@ const drwMutexAcquireTimeout = 1 * time.Second // 1 second.
 const drwMutexRefreshCallTimeout = 5 * time.Second
 
 // dRWMutexUnlockTimeout - timeout for the unlock call
-const drwMutexUnlockCallTimeout = 20 * time.Second
+const drwMutexUnlockCallTimeout = 30 * time.Second
 
 // dRWMutexRefreshInterval - the interval between two refresh calls
 const drwMutexRefreshInterval = 10 * time.Second

--- a/pkg/dsync/drwmutex_test.go
+++ b/pkg/dsync/drwmutex_test.go
@@ -15,7 +15,7 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-package dsync_test
+package dsync
 
 import (
 	"context"
@@ -24,8 +24,6 @@ import (
 	"sync/atomic"
 	"testing"
 	"time"
-
-	. "github.com/minio/minio/pkg/dsync"
 )
 
 const (

--- a/pkg/dsync/dsync-server_test.go
+++ b/pkg/dsync/dsync-server_test.go
@@ -15,13 +15,13 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-package dsync_test
+package dsync
 
 import (
 	"fmt"
 	"sync"
-
-	. "github.com/minio/minio/pkg/dsync"
+	"sync/atomic"
+	"time"
 )
 
 const WriteLock = -1
@@ -34,6 +34,9 @@ type lockServer struct {
 
 	// Refresh returns lock not found if set to true
 	lockNotFound bool
+
+	// Set to true if you want peers servers to do not respond
+	responseDelay int64
 }
 
 func (l *lockServer) setRefreshReply(refreshed bool) {
@@ -42,7 +45,15 @@ func (l *lockServer) setRefreshReply(refreshed bool) {
 	l.lockNotFound = !refreshed
 }
 
+func (l *lockServer) setResponseDelay(responseDelay time.Duration) {
+	atomic.StoreInt64(&l.responseDelay, int64(responseDelay))
+}
+
 func (l *lockServer) Lock(args *LockArgs, reply *bool) error {
+	if d := atomic.LoadInt64(&l.responseDelay); d != 0 {
+		time.Sleep(time.Duration(d))
+	}
+
 	l.mutex.Lock()
 	defer l.mutex.Unlock()
 	if _, *reply = l.lockMap[args.Resources[0]]; !*reply {
@@ -53,6 +64,10 @@ func (l *lockServer) Lock(args *LockArgs, reply *bool) error {
 }
 
 func (l *lockServer) Unlock(args *LockArgs, reply *bool) error {
+	if d := atomic.LoadInt64(&l.responseDelay); d != 0 {
+		time.Sleep(time.Duration(d))
+	}
+
 	l.mutex.Lock()
 	defer l.mutex.Unlock()
 	var locksHeld int64
@@ -69,6 +84,10 @@ func (l *lockServer) Unlock(args *LockArgs, reply *bool) error {
 const ReadLock = 1
 
 func (l *lockServer) RLock(args *LockArgs, reply *bool) error {
+	if d := atomic.LoadInt64(&l.responseDelay); d != 0 {
+		time.Sleep(time.Duration(d))
+	}
+
 	l.mutex.Lock()
 	defer l.mutex.Unlock()
 	var locksHeld int64
@@ -84,6 +103,10 @@ func (l *lockServer) RLock(args *LockArgs, reply *bool) error {
 }
 
 func (l *lockServer) RUnlock(args *LockArgs, reply *bool) error {
+	if d := atomic.LoadInt64(&l.responseDelay); d != 0 {
+		time.Sleep(time.Duration(d))
+	}
+
 	l.mutex.Lock()
 	defer l.mutex.Unlock()
 	var locksHeld int64
@@ -102,6 +125,10 @@ func (l *lockServer) RUnlock(args *LockArgs, reply *bool) error {
 }
 
 func (l *lockServer) Refresh(args *LockArgs, reply *bool) error {
+	if d := atomic.LoadInt64(&l.responseDelay); d != 0 {
+		time.Sleep(time.Duration(d))
+	}
+
 	l.mutex.Lock()
 	defer l.mutex.Unlock()
 	*reply = !l.lockNotFound
@@ -109,6 +136,10 @@ func (l *lockServer) Refresh(args *LockArgs, reply *bool) error {
 }
 
 func (l *lockServer) ForceUnlock(args *LockArgs, reply *bool) error {
+	if d := atomic.LoadInt64(&l.responseDelay); d != 0 {
+		time.Sleep(time.Duration(d))
+	}
+
 	l.mutex.Lock()
 	defer l.mutex.Unlock()
 	if len(args.UID) != 0 {

--- a/pkg/dsync/dsync_test.go
+++ b/pkg/dsync/dsync_test.go
@@ -15,12 +15,12 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-package dsync_test
+package dsync
 
 import (
 	"context"
 	"fmt"
-	"log"
+	golog "log"
 	"math/rand"
 	"net"
 	"net/http"
@@ -32,8 +32,6 @@ import (
 	"time"
 
 	"github.com/google/uuid"
-	"github.com/minio/minio/pkg/dsync"
-	. "github.com/minio/minio/pkg/dsync"
 )
 
 const numberOfNodes = 5
@@ -56,7 +54,7 @@ func startRPCServers() {
 		server.HandleHTTP(rpcPaths[i], fmt.Sprintf("%s-debug", rpcPaths[i]))
 		l, e := net.Listen("tcp", ":"+strconv.Itoa(i+12345))
 		if e != nil {
-			log.Fatal("listen error:", e)
+			golog.Fatal("listen error:", e)
 		}
 		go http.Serve(l, nil)
 
@@ -244,6 +242,7 @@ func TestFailedRefreshLock(t *testing.T) {
 	// Simulate Refresh RPC response to return no locking found
 	for i := range lockServers {
 		lockServers[i].setRefreshReply(false)
+		defer lockServers[i].setRefreshReply(true)
 	}
 
 	dm := NewDRWMutex(ds, "aap")
@@ -256,7 +255,7 @@ func TestFailedRefreshLock(t *testing.T) {
 		wg.Done()
 	}
 
-	if !dm.GetLock(ctx, cancel, id, source, dsync.Options{Timeout: 5 * time.Minute}) {
+	if !dm.GetLock(ctx, cancel, id, source, Options{Timeout: 5 * time.Minute}) {
 		t.Fatal("GetLock() should be successful")
 	}
 
@@ -268,10 +267,35 @@ func TestFailedRefreshLock(t *testing.T) {
 
 	// Should be safe operation in all cases
 	dm.Unlock()
+}
 
-	// Revert Refresh RPC response to locking found
+// Test Unlock should not timeout
+func TestUnlockShouldNotTimeout(t *testing.T) {
+	dm := NewDRWMutex(ds, "aap")
+
+	if !dm.GetLock(context.Background(), nil, id, source, Options{Timeout: 5 * time.Minute}) {
+		t.Fatal("GetLock() should be successful")
+	}
+
+	// Add delay to lock server responses to ensure that lock does not timeout
 	for i := range lockServers {
-		lockServers[i].setRefreshReply(false)
+		lockServers[i].setResponseDelay(2 * drwMutexUnlockCallTimeout)
+		defer lockServers[i].setResponseDelay(0)
+	}
+
+	unlockReturned := make(chan struct{}, 1)
+	go func() {
+		dm.Unlock()
+		unlockReturned <- struct{}{}
+	}()
+
+	timer := time.NewTimer(2 * drwMutexUnlockCallTimeout)
+	defer timer.Stop()
+
+	select {
+	case <-unlockReturned:
+		t.Fatal("Unlock timed out, which should not happen")
+	case <-timer.C:
 	}
 }
 

--- a/pkg/dsync/rpc-client-impl_test.go
+++ b/pkg/dsync/rpc-client-impl_test.go
@@ -15,14 +15,12 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-package dsync_test
+package dsync
 
 import (
 	"context"
 	"net/rpc"
 	"sync"
-
-	. "github.com/minio/minio/pkg/dsync"
 )
 
 // ReconnectRPCClient is a wrapper type for rpc.Client which provides reconnect on first failure.

--- a/pkg/dsync/rpc-client-interface.go
+++ b/pkg/dsync/rpc-client-interface.go
@@ -54,11 +54,15 @@ type NetLocker interface {
 	// Do read unlock for given LockArgs. It should return
 	// * a boolean to indicate success/failure of the operation
 	// * an error on failure of unlock request operation.
+	// Canceling the context will abort the remote call.
+	// In that case, the resource may or may not be unlocked.
 	RUnlock(ctx context.Context, args LockArgs) (bool, error)
 
 	// Do write unlock for given LockArgs. It should return
 	// * a boolean to indicate success/failure of the operation
 	// * an error on failure of unlock request operation.
+	// Canceling the context will abort the remote call.
+	// In that case, the resource may or may not be unlocked.
 	Unlock(ctx context.Context, args LockArgs) (bool, error)
 
 	// Refresh the given lock to prevent it from becoming stale

